### PR TITLE
Update conan install tasks in gradle (allows running ./gradlew clean without running conan install)

### PR DESCRIPTION
**Pull request by ViliusSutkus89**
Hello,

I've updated conan tasks in [build.gradle](app/build.gradle). Now they only execute only when invoked. Gradle tasks are not exactly functions, the body of each task is executed when loading gradle file. Tasks can also contain `doFirst` and `doLast` which are executed only when the task is invoked, not just loaded. Changed `conanProfile` task type to `Copy`, which takes `from` and `into` arguments and executes them in it's `doFirst` stage. Can't change `conanInstall` task type to Exec, because Exec type usually only executes once.

Made the `conanInstall` task as a dependency of task `preBuild` and `conanProfile` task as a dependency of task `conanInstall`.

Long story short - this allows running `./gradlew clean` without executing those tasks.

As a bonus, I've added additional task to clean `.cxx` directory during `gradle clean`. `.cxx` is an additional build directory, no idea why regular `gradle clean` is not cleaning it up.

### CODE REVIEW

**Comment by ViliusSutkus89** [comment]
ok, so I've made a mistake of creating a MR from a repo in my own profile, not in the organization. This means that the CI is failing because of failed login to artifactory. Moved artifactory login step a bit later - right before conan upload. Removed login step altogether from emulator job. Emulators should succeed, build job will fail, but successful gradle run is visible in the workflow run.

**Comment by ViliusSutkus89** [comment]
Had to add another workaround. Since we supply our own `CMAKE_TOOLCHAIN_FILE`,
Android Gradle Plugin expects toolchain file to be there. If toolchain file is missing - `gradle clean` task fails.
Toolchain files are generated by `conan install`, but if something fails before ALL of the toolchain files are created - `gradle clean` will fail too. It's not a nice state where a failed build causes cleanup task to fail too, indefinitely.

The workaround checks if `conan_toolchain.cmake` files are there, and if they are missing - removes gradle `clean` task dependency on tasks `externalNativeBuildCleanLiteDebug`, `externalNativeBuildCleanLiteRelease`, `...`.

File: .github/workflows/android_main.yml [hunk 1]
```diff
@@ -118,8 +119,6 @@ jobs:
 
     - name: conan remote
       run: conan remote add --index 0 odr https://artifactory.opendocument.app/artifactory/api/conan/conan
-    - name: conan login
```
Comment (TomTasche) [inline .github/workflows/android_main.yml:121]: Why was that deleted? On line 175 we do upload some binaries...

File: app/build.gradle [hunk 2]
```diff
@@ -8,21 +8,21 @@ android {
...
-    copy {
-        from "conanprofile.txt"
-        into "build"
-    }
+tasks.register('conanProfile', Copy) {
+    from "conanprofile.txt"
+    into "build/"
 
+    doLast {
     def file = file("build/conanprofile.txt")
```
Comment (TomTasche) [inline app/build.gradle:16]: Increase Indentation?

File: app/build.gradle [hunk 3]
```diff
@@ -8,21 +8,21 @@ android {
...
     content = content.replace("<NDK_PATH>", android.ndkDirectory.toString())
     file.write(content)
+    }
 }
 
-task conanInstall {
+tasks.register('conanInstall') {
+    dependsOn(conanProfile)
+    doFirst {
     ["armv7", "armv8", "x86", "x86_64"].each { String arch ->
```
Comment (TomTasche) [inline app/build.gradle:26]: Indentation?

File: app/build.gradle [hunk 4]
```diff
@@ -168,3 +170,31 @@ dependencies {
...
+// This is needed because task "clean" depends on these tasks:
+//:app:externalNativeBuildCleanLiteDebug
+//:app:externalNativeBuildCleanLiteRelease
+//:app:externalNativeBuildCleanProDebug
+//:app:externalNativeBuildCleanProRelease
+// And these tasks expect to find build/conan/[ABI]/conan_toolchain.cmake files
+//
+// If build folder is malformed, we can't run `gradle clean` without this workaround.
+// This is because we supply our (conan's) CMAKE_TOOLCHAIN_FILE.
+// This workaround removes these "clean" task dependencies if conan_toolchain.cmake is not found (malformed build dir)
```
Comment (TomTasche) [inline app/build.gradle:189]: How about creating an empty file at build/conan/[ABI]/conan_toolchain.cmake instead to make sure those clean-tasks execute successfully? That way we don't lose any additional functionality of those tasks (if there is any, ever). Seems a bit less hacky to me that way!

**Review by TomTasche — COMMENTED** [review]
Nice!

**Review by andiwand — COMMENTED** [review]
I think ultimately we should move this out of gradle again and use a script like we do here https://github.com/opendocument-app/OpenDocument.ios/blob/main/conan/setup-all.sh

**Comment by TomTasche** [comment]
> I think ultimately we should move this out of gradle again and use a script like we do here https://github.com/opendocument-app/OpenDocument.ios/blob/main/conan/setup-all.sh

Didn't you spend a lot of time on getting the native build into gradle in the first place? I don't see why that's a bad thing. One less step to worry about during setup!

**Comment by andiwand** [comment]
Maybe but I don't think it is a great approach because of changing the PATH and coupling two independent tools together which provide zero integration. But that is just my preference. If you two thing it is fine as is I will not change it

File: .github/workflows/android_main.yml (hunk 1 above)
Comment (ViliusSutkus89) [reply .github/workflows/android_main.yml:121]: Somehow I've missed that part. I was assuming we only upload binaries in build, not test jobs. fixed now.

File: app/build.gradle (hunk 4 above)
Comment (ViliusSutkus89) [reply app/build.gradle]: Tried that, couldn't figure out why it doesn't work properly, CMake expects more files. Anyway another simpler workaround already in place

**Comment by ViliusSutkus89** [comment]
I've changed `gradle clean` workaround to just removing `.cxx` dir in `doFirst` stage. It's needed to make `./gradlew clean; ./gradlew clean` work.

`conanInstall` creates toolchain and conan dependency module files somewhere in build dir. Some config in .cxx references them. `./gradlew clean` removes build dir, but keeps .cxx dir. Since .cxx dir still there, another run of `./gradlew clean` tries to look up those files and fails. Long story short, just removing .cxx dir allows straight up removing build dir without worrying about referencing anything and `./gradlew clean; ./gradlew clean` just works.

Without this whole MR, `./gradlew clean; ./gradlew clean` would also work, but just because conan install is launched on each `./gradlew clean`, so there's always something to clean up. Since this MR prevents `conan install` from being runned on each gradle invocation, this issue becomes apparent

**Review by TomTasche — APPROVED** [review]
Looks good.

**Comment by ViliusSutkus89** [comment]
### External script vs gradle task

We could rephrase the PATH requirement into a requirement of having conan installed. I mean properly installed. I think it's normal to require some build tools. In CI we install it with `pip install conan` and that's that.

Solving this issue as a gradle task makes usage easier, because you can just open Android Studio and usually all is handled. With external script, we need to know (and also document) extra steps that need to be done before Android Studio can be even opened. In conanInstall task we could check if conan is available and if not, print a nice error message suggesting `pip install`. Or `pip install` to build dir dir and then use it to do `conan install`

**Comment by andiwand** [comment]
I think for example Ubuntu does not allow you to install packages in system or user space anymore but only via apt or via venv which makes the `PATH` approach a bit tricky. But at the end it depends on us and how we want to use it

**Comment by ViliusSutkus89** [comment]
`setup-all.sh` would also be affected by that, right?

**Comment by andiwand** [comment]
Right but then you get a change to source your venv while with Android Studio you have to populate the venv somewhere more global

**Comment by ViliusSutkus89** [comment]
We don't need the whole Android Studio to consume the venv. In `conanInstall` task we can have a wrapper, which calls `conan install` in a venv

**Comment by ViliusSutkus89** [comment]
Update: preBuild depending on conanInstall means that build order is fine, but turns out that Android Studio sync doesn't call preBuild, thus doesn't execute conanInstall, thus throws a sync error when trying to sync an unbuilt project. Will figure it out

**Comment by ViliusSutkus89** [comment]
Android Studio's sync should work now on empty build tree too

**Review by TomTasche — COMMENTED** [review]
Ready to merge from my side. Thanks for testing all the edge cases!

**Review by andiwand — APPROVED** [review]
LGTM

**Comment by TomTasche** [comment]
> I think for example Ubuntu does not allow you to install packages in system or user space anymore but only via apt or via venv which makes the `PATH` approach a bit tricky. But at the end it depends on us and how we want to use it

Just leaving this here because I will definitely stumble upon it again in the future. Android Studio on Ubuntu does not respect the user's PATH environment variable: https://stackoverflow.com/a/69419653/198996
